src/Loader.php

```diff
@@ -479,6 +479,140 @@ public static function register_scripts() {
+
+		return $filename;
+	}
+
+	/**
+	 * Find and combine translation chunk files.
+	 *
+	 * Only targets files that aren't represented by a registered script (e.g. not passed to wp_register_script()).
+	 *
+	 * @param string $lang_dir Path to language files.
+	 * @param string $domain Text domain.
+	 * @param string $locale Locale being retrieved.
+	 * @return array Combined translation chunk data.
+	 */
+	public static function get_translation_chunk_data( $lang_dir, $domain, $locale ) {
+		global $wp_filesystem;
+
+		// Grab all JSON files in the current language pack.
+		$json_i18n_filenames       = glob( $lang_dir . $domain . '-' . $locale . '-*.json' );
+		$combined_translation_data = array();
+
+		foreach ( $json_i18n_filenames as $json_filename ) {
+			if ( ! $wp_filesystem->is_readable( $json_filename ) ) {
+				continue;
+			}
+
+			$file_contents = $wp_filesystem->get_contents( $json_filename );
+			$chunk_data    = \json_decode( $file_contents, true );
+
+			if ( empty( $chunk_data ) ) {
+				continue;
+			}
+
+			if ( ! isset( $chunk_data['comment']['reference'] ) ) {
+				continue;
+			}
+
+			$reference_file = $chunk_data['comment']['reference'];
+
+			// Only combine "app" files (not scripts registered with WP).
+			if (
+				false === strpos( $reference_file, 'dist/chunks/' ) &&
+				false === strpos( $reference_file, 'dist/app/index.js' )
+			) {
+				continue;
+			}
+
+			if ( empty( $combined_translation_data ) ) {
+				// Use the first translation file as the base structure.
+				$combined_translation_data = $chunk_data;
+			} else {
+				// Combine all messages from all chunk files.
+				$combined_translation_data['locale_data']['messages'] = array_merge(
+					$combined_translation_data['locale_data']['messages'],
+					$chunk_data['locale_data']['messages']
+				);
+			}
+		}
+
+		// Remove inaccurate reference comment.
+		unset( $combined_translation_data['comment'] );
+
+		return $combined_translation_data;
+	}
+
+	/**
+	 * Load translation strings from language packs for dynamic imports.
+	 *
+	 * This function combines JSON translation data auto-extracted by GlotPress
+	 * from Webpack-generated JS chunks into a single file that can be used in
+	 * subsequent requests. This is necessary since the JS chunks are not known
+	 * to WordPress via wp_register_script() and wp_set_script_translations().
+	 *
+	 * @param string $original_translations JSON encoded translations object.
+	 * @param string $file File location for the script being translated.
+	 * @param string $handle Script handle.
+	 * @param string $domain Text domain.
+	 *
+	 * @return string JSON encoded translations object.
+	 */
+	public static function load_script_translations( $original_translations, $file, $handle, $domain ) {
+		// Make sure the main app script is being loaded.
+		if ( WC_ADMIN_APP !== $handle ) {
+			return $original_translations;
+		}
+
+		// Make sure we're handing the correct domain (could be woocommerce or woocommerce-admin).
```

Comment:
If we are already verifying that the script is `WC_ADMIN_APP` would there ever be a chance that the text domain would be different? I'm guessing it would only be if another plugin used the same file/handle name?

Reply:
Yeah this is definitely overly defensive. Another plugin would need to have a `wc-admin-app` script.